core/runtime.rs

```diff
@@ -75,31 +85,134 @@ struct IsolateAllocations {
+  /// happen in tests (and possibly for users using the inspector).
+  pub fn prepare_for_cleanup(&mut self) {
+    let mut state = self.state.borrow_mut();
+    let inspector = state.inspector.take();
+    state.op_state.borrow_mut().clear();
+    if let Some(inspector) = inspector {
+      assert_eq!(
+        Rc::strong_count(&inspector),
+        1,
+        "The inspector must be dropped before the runtime"
+      );
+    }
+  }
+
+  pub fn cleanup(&mut self) {
+    self.prepare_for_cleanup();
+
+    let state_ptr = self.v8_isolate.get_data(STATE_DATA_OFFSET);
+    // SAFETY: We are sure that it's a valid pointer for whole lifetime of
+    // the runtime.
+    _ = unsafe { Rc::from_raw(state_ptr as *const RefCell<JsRuntimeState>) };
+
+    let module_map_ptr = self.v8_isolate.get_data(MODULE_MAP_DATA_OFFSET);
+    // SAFETY: We are sure that it's a valid pointer for whole lifetime of
+    // the runtime.
+    _ = unsafe { Rc::from_raw(module_map_ptr as *const RefCell<ModuleMap>) };
+
+    self.state.borrow_mut().destroy_all_realms();
+
+    debug_assert_eq!(Rc::strong_count(&self.state), 1);
+  }
+
+  pub fn prepare_for_snapshot(mut self) -> v8::OwnedIsolate {
+    self.cleanup();
+    // SAFETY: We're copying out of self and then immediately forgetting self
+    let (state, isolate) = unsafe {
+      (
+        ManuallyDrop::take(&mut self.state.0),
+        ManuallyDrop::take(&mut self.v8_isolate),
+      )
+    };
+    std::mem::forget(self);
+    drop(state);
+    isolate
+  }
+}
+
+impl Drop for InnerIsolateState {
+  fn drop(&mut self) {
+    self.cleanup();
+    // SAFETY: We gotta drop these
+    unsafe {
+      ManuallyDrop::drop(&mut self.state.0);
+      if self.snapshotting {
+        // Create the snapshot and just drop it.
+        eprintln!("WARNING: v8::OwnedIsolate for snapshot was leaked");
+      } else {
+        ManuallyDrop::drop(&mut self.v8_isolate);
+      }
+    }
+  }
+}
+
 /// A single execution context of JavaScript. Corresponds roughly to the "Web
-/// Worker" concept in the DOM. A JsRuntime is a Future that can be used with
+/// Worker" concept in the DOM. A JsRuntimeImpl is a Future that can be used with
 /// an event loop (Tokio, async_std).
 ////
-/// The JsRuntime future completes when there is an error or when all
+/// The JsRuntimeImpl future completes when there is an error or when all
 /// pending ops have completed.
 ///
 /// Pending ops are created in JavaScript by calling Deno.core.opAsync(), and in Rust
 /// by implementing an async function that takes a serde::Deserialize "control argument"
 /// and an optional zero copy buffer, each async Op is tied to a Promise in JavaScript.
-pub struct JsRuntime {
-  state: Rc<RefCell<JsRuntimeState>>,
+///
+/// API consumers will want to use either the [`JsRuntime`] or [`JsRuntimeForSnapshot`]
+/// type aliases.
+pub struct JsRuntimeImpl<const FOR_SNAPSHOT: bool = false> {
```

Comment:
I'm a bit wary of this struct - would it be possible to have `JsRuntimeForSnapshot` that wraps `JsRuntime` instead and derefs to it?

Reply:
I can do that. The type aliases work the same way but it's easy enough.

Reply:
That would be great, thanks!

Reply:
Right now we have
```rust
pub struct JsRuntime(JsRuntimeImpl<false>);
pub struct JsRuntimeForSnapshot(JsRuntimeImpl<true>);
```
but I think what was meant by this review was
```rust
pub struct JsRuntime {
  // ...
}

pub struct JsRuntimeForSnapshot {
  inner: JsRuntime
  // ...
}
```

Reply:
@nayeemrmn  That makes sense and probably results in a cleaner API experience -- we can definitely do that (and if you want to tackle it I can review it, especially while it's fresh in my mind).